### PR TITLE
Change language code from fi_FI to fi

### LINKED ISSUE #869: Too specific locale fi_FI
### Steps to reproduce
1. Open Firefox or Chrome, configure Finnish as preferred language
2. Remove config/config.php default_language.
3. Share a link, open the link in Firefox or Chrome
### Expected behaviour

Should see the link page in Finnish language.
### Actual behaviour

Shows the link page in English.

Browser sends "Accept-Language: fi,en-US;q=0.7,en;q=0.3". The undefined default_language lets Nextcloud determine language from this request header. Nextcloud has language fi_FI, but no fi. Therefore Nextcloud defaults to en. Internet Explorer 11 has option to select both fi and fi-FI language, where fi-FI correctly selects the Nextcloud fi_FI language. Firefox and Chrome allow only fi language. Also, if Nextcould had language fi, but no fi_FI, I think Accept-Language of fi_FI would fall back to fi (see OC\L10N\Factory::findAvailableLanguages in lib/lib/private/l10n/factory.php). Therefore I think it is wrong that Nextcloud has specific fi_FI instead of the less specific fi. The same issue seems to be with other languages as well (af, am, bg, cs, cy, fy, hu, ka, ku, ms, mt, my, nb, nn, sw, tg, th, tl, ur, by looking at core/l10n), and I'm not sure if every language should have a generic fallback.
### Server configuration

**Operating system**: Linux, openSUSE 13.2

**Web server:** Apache

**Database:** MySQL

**PHP version:** 5.6.1

**Nextcloud version:** Nextcloud 9.0.53 (stable)

**Updated from an older Nextcloud/ownCloud or fresh install:** fresh

**Where did you install Nextcloud from:** nextcloud.com nextcloud-9.0.53.tar.bz2

**Signing status:**
No errors have been found.

**List of activated apps:**
Enabled:
- activity: 2.2.1
- admin_audit: 1.0.0
- comments: 0.2
- dav: 0.1.6
- federatedfilesharing: 0.1.0
- federation: 0.0.4
- files: 1.4.4
- files_pdfviewer: 0.8.1
- files_sharing: 0.9.1
- files_texteditor: 2.1
- files_trashbin: 0.8.0
- files_versions: 1.2.0
- files_videoplayer: 0.9.8
- firstrunwizard: 1.1
- gallery: 14.5.0
- notifications: 0.2.3
- password_policy: 1.0.0
- provisioning_api: 0.4.1
- systemtags: 0.2
- templateeditor: 0.1
- theming: 0.1.0
- updatenotification: 0.1.0
  Disabled:
- encryption
- external
- files_external
- user_external
- user_ldap
- user_saml

**The content of config/config.php:**
{
    "system": {
        "passwordsalt": "**_REMOVED SENSITIVE VALUE**_",
        "secret": "**_REMOVED SENSITIVE VALUE**_",
        "trusted_domains": [
            "**_REMOVED SENSITIVE VALUE**_"
        ],
        "datadirectory": "**_REMOVED SENSITIVE VALUE**_",
        "overwrite.cli.url": "**_REMOVED SENSITIVE VALUE**_",
        "htaccess.RewriteBase": "\/",
        "dbtype": "mysql",
        "version": "9.0.53.0",
        "dbname": "**_REMOVED SENSITIVE VALUE**_",
        "dbhost": "localhost",
        "dbtableprefix": "oc_",
        "dbuser": "**_REMOVED SENSITIVE VALUE**_",
        "dbpassword": "**_REMOVED SENSITIVE VALUE**_",
        "logtimezone": "UTC",
        "installed": true,
        "instanceid": "**_REMOVED SENSITIVE VALUE**_",
        "singleuser": false,
        "skeletondirectory": "**_REMOVED SENSITIVE VALUE**_",
        "mail_from_address": "noreply",
        "mail_smtpmode": "php",
        "mail_domain": "**_REMOVED SENSITIVE VALUE**_"
    }
}

**Are you using external storage, if yes which one:** no

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** no
### Client configuration

**Browser:** Firefox 48.0 or Chrome 52.0.2743.116 m

**Operating system:** Windows 7 Pro
#### Web server error log

Empty
#### Nextcloud log (data/nextcloud.log)

Unrelated errors
Error   PHP     Undefined variable: share at .../lib/private/share20/manager.php#996
Error   PHP Undefined index: id at .../core/ajax/share.php#229
Error   PHP Cannot modify header information - headers already sent by (output started at .../3rdparty/sabre/http/lib/Sapi.php:83) at .../3rdparty/sabre/http/lib/Sapi.php#63

### PR DESCRIPTION
**Pull request by MorrisJobke**
- then the language is not that specific and get also matched for fi
- fallback from fi_FI to fi is supported - the other way around not
- contains repair script
- contains tests for repair script
- fixes #869 

@kohtala @rullzer 

I will also update all the transifex files in all apps to match the settings here.

### CODE REVIEW

**Comment by rullzer** [comment]
* Updated autoloader.

:+1:

**Comment by kohtala** [comment]
Hi. Thanks for working on this.

In the #869 I had listed 18th Aug that this same problem seems to be with several other languages as well. I prepared the list for other languages for lang_map as af_zA:af, am_ET:am, bg_BG:bg, cs_CZ:cs, cy_GB:cy, fy_NL:fy, hu_HU:hu, ka_GE:ka, ku_IQ:ku, lt_LT:lt, ms_MY:ms, mt_MT:mt, my_MM:my, nb_NO:nb, nn_NO:nn, si_LK:si, sk_SK:sk, sw_KE:sw, tg_TJ:tg, th_TH:th, tl_PH:tl, ur_PK:ur.

Also, would you think /l10n/rm-old.sh, /settings/languageCodes.php and /settings/personal.php need an update as well?

**Comment by MorrisJobke** [comment]
> Also, would you think /l10n/rm-old.sh, /settings/languageCodes.php and /settings/personal.php need an update as well?

Thanks. I will check this.

File: l10n/.tx/config [hunk 1]
```diff
@@ -1,6 +1,6 @@
 [main]
 host = https://www.transifex.com
-lang_map = ja_JP: ja
+lang_map = ja_JP: ja, fi_FI: fi
```
Comment (nickvergessen) [inline l10n/.tx/config]: does every app need this change? or is it good enough, once per project?

Comment (MorrisJobke) [inline l10n/.tx/config]: Every app needs this, but I will do that once this is in.

**Comment by LukasReschke** [comment]
@MorrisJobke Is this still interesting? If not we should probably close this 😉

**Comment by MorrisJobke** [comment]
> @MorrisJobke Is this still interesting? If not we should probably close this 😉

Yes. One needs to write the migration code to change the users language to fi if it was fi_FI.

**Comment by LukasReschke** [comment]
> Yes. **One** needs to write the migration code to change the users language to fi if it was fi_FI.

morris ❓

**Comment by blizzz** [comment]
Can be mostly guttenberged from DAV migration, see https://github.com/nextcloud/server/blob/master/apps/dav/lib/Migration/ValueFixInsert.php and https://github.com/nextcloud/server/blob/master/apps/dav/lib/Migration/ValueFix.php

Basically it registers a background job per (known) user and applies some logic and the actual implementation.

**Comment by nickvergessen** [comment]
Or simply run a query `update oc_preferences set value =“fi“ where value = “fi_FI“ and key =“lang“ and app =“settings“` ? No need to loop over users or create a background job....

**Comment by blizzz** [comment]
@nickvergessen ah, of course, this is most reasonably :)

**Comment by MorrisJobke** [comment]
Ready for review.

**Review by nickvergessen — APPROVED** [review]
Extended the tests for a case with more than 1 user and fixed the table name in the output

**Review by rullzer — APPROVED** [review]
Looks good to me :+1:

**Comment by MorrisJobke** [comment]
I need to look into fixing the backports of the wrong translations or we simply backport this one here. https://github.com/nextcloud/server/commit/2282e61e2ec58c2baffee48acab63228b03c1223

Maybe I just copy over the now renamed stuff to the old version for all the changed languages in the translation script.

**Pull request merged**